File: 알고리즘/전승민/wordTranslate/wordTranslate.kt
import kotlin.math.min

fun wordTranslate(begin: String, target: String, words: Array<String>): Int
    = dps(begin, target, Array(words.size) { false }, words).let {
        if(it==infinity) 0 else it
    }

fun dps(cur: String, target: String, visit: Array<Boolean>, words: Array<String>): Int {
    if (cur == target) {
        return 0
    }
    var minimum = infinity
    words.indices.forEach {
        if (!visit[it]) {
            visit[it] = true
            minimum = if (cur.isOneLetterDiff(words[it])) min(dps(words[it], target, visit, words) + 1, minimum) else minimum
            visit[it] = false
        }
    }
    return minimum
}

fun String.isOneLetterDiff(other: String): Boolean {
    var count = 0
    this.indices.forEach {
        if (this[it] == other[it]) count++
    }
    return this.length - 1 == count
}
// https://programmers.co.kr/learn/courses/30/lessons/43163

File: 알고리즘/김준우/wordTranslate.js
function solution(begin, target, words) {
  if (!words.includes(target)) return 0;
  const result = [];

  function bfs(begin, words, count) {
    const queue = [];

    words.forEach((word, index) => {
      const wordCount = word
        .split("")
        .reduce((acc, v, i) => (begin[i] !== v ? acc + 1 : acc), 0);

      if (wordCount === 1) {
        if (word === target) {
          result.push(count + 1);
          return;
        }

        const dummy = [...words];
        dummy.splice(index, 1);
        queue.push({ word, dummy });
      }
    });

    if (queue.length) {
      ++count;
      queue.forEach((v) => {
        bfs(v.word, v.dummy, count);
      });
    }
  }

  bfs(begin, words, 0);

  return result.sort((a, b) => a - b)[0];
}

// https://programmers.co.kr/learn/courses/30/lessons/43163
